Fix Zbus Benchmark Sample is smashing the stack and is too fast

alpsayin opened the pull request:
ZBus subscriber threads were smashing their threads. Upped their stack with something safe by utilising `IDLE_STACK_SIZE` for subscribers and `MAIN_STACK_SIZE` for producer.

Benchmark also finishes way too fast for most if not all targets. Used cycles instead of uptime as originally suggested by @rodrigopex and used nanosecond conversions to get around the `duration==0` check.

Native posix still returns `start` and `end` times as `0` though, so with further commits utilised `native_rtc` with `RTC_CLOCK_PSEUDOHOSTREALTIME` to get as much resolution as possible.

alpsayin committed:
samples: subsys: zbus: benchmark: increase thread stack sizes

Stack sizes need to be dependent on something arch specific such as
idle stack size because they're smashing their stacks.

Signed-off-by: Alp Sayin <alp.sayin@amd.com>

alpsayin committed:
samples: subsys: zbus: benchmark: remove all platform exclusions

The sample is now able to run on all previously excluded platforms.

Signed-off-by: Alp Sayin <alp.sayin@amd.com>

asayin-xlnx commented:
@rodrigopex for reviews?

alpsayin commented:
@rodrigopex for reviews?

rodrigopex commented on samples/subsys/zbus/benchmark/src/benchmark.c:
```diff
@@ -197,18 +197,21 @@ static void producer_thread(void)
 
 	if (duration == 0) {
 		LOG_ERR("Something wrong. Duration is zero!\n");
+		LOG_ERR("start = %" PRIu32 "\n", start);
+		LOG_ERR("end = %" PRIu32 "\n", end);
```
These two lines (205 and 206) are unnecessary. Line 204 is telling that already.

asayin-xlnx replied:
Done!

rodrigopex commented on samples/subsys/zbus/benchmark/src/benchmark.c:
```diff
@@ -197,18 +197,21 @@ static void producer_thread(void)
 	}
-	uint64_t i = (BYTES_TO_BE_SENT * 1000) / duration;
-	uint64_t f = ((BYTES_TO_BE_SENT * 100000) / duration) % 100;
+	uint64_t i = (BYTES_TO_BE_SENT * NSEC_PER_SEC / MB(1)) / duration;
+	uint64_t f = (BYTES_TO_BE_SENT * NSEC_PER_SEC * 100 / MB(1) / duration) % 100;
```
I would add parenthesis around `BYTES_TO_BE_SENT * NSEC_PER_SEC` and `BYTES_TO_BE_SENT * NSEC_PER_SEC * 100` just to make explicit the precedence I desire.

asayin-xlnx replied:
Done!

rodrigopex reviewed (changes requested):
Only small changes are necessary.

rodrigopex commented:
@alpsayin maybe we could leave the duration zero occurs. The only thing we need to do is to avoid the zero division and force `f`and `i` also to be zero. What do you think?

alpsayin committed:
samples: subsys: zbus: benchmark: use sys_clock_cycle instead of k_uptime

Needed because benchmark finishes faster than a tick, especially on
emulated platforms where tick rate is 100Hz. Output units are also updated
to print numbers with less digits.

Signed-off-by: Alp Sayin <alp.sayin@amd.com>

asayin-xlnx commented:
> @alpsayin maybe we could leave the duration zero occurs. The only thing we need to do is to avoid the zero division and force `f`and `i` also to be zero. What do you think? And that would remove `platform_exclude: native_posix` from the sample yaml you are suggesting.

I found a somewhat reliable way to measure time for `native_posix`. "Reliable" is when you up the `BYTES_TO_BE_SENT` to MBs then duration becomes consistently directly proportional _(as one would expect)_ to `BYTES_TO_BE_SENT`. With 256kB, the deviation is about 30% on the machine that I'm testing (workstation with 0 load).

Eyeballed means/deviations are 30+/-10us for `native_posix_64` and 35+/-5us for `native_posix`.

Let's see what CI/CD will end up showing.

rodrigopex commented on samples/subsys/zbus/benchmark/src/benchmark.c:
```diff
@@ -11,8 +11,16 @@
+
 LOG_MODULE_DECLARE(zbus, CONFIG_ZBUS_LOG_LEVEL);
 
+#define CNS_STACK_SIZE (CONFIG_IDLE_STACK_SIZE + CONFIG_BM_MESSAGE_SIZE)
+#define PRD_STACK_SIZE (CONFIG_MAIN_STACK_SIZE + CONFIG_BM_MESSAGE_SIZE)
```
I would use the `CONSUMER_STACK_SIZE` and `PRODUCER_STACK_SIZE` names instead of `CNS_STACK_SIZE` and `PRD_STACK_SIZE.` There is no bonus for using the abbreviated ones.

alpsayin replied:
Done! 
This is a stupid one. Yours was my first intention too. 
Then `K_THREAD_DEFINE` lines got too long and `checkpatch` complained. 
I broke the lines and `checkpatch` complained about "unnecessarily broken line". 
I got mad and butchered the macro name... 🤦🏼

rodrigopex replied:
Have you tried the check patch “--fix-inplace”?

alpsayin replied:
Yes, unfortunately, `--fix` nor `--fix-inplace` did nothing for this specific one. 
Eventually `checkpatch` also said trust your own judgement, and so I decided not to break this last line since it's only 102 chars long. `¯\_(ツ)_/¯`

https://github.com/zephyrproject-rtos/zephyr/pull/53563/commits/3edb17edb648fd3c587f2cc8a332c660425153f9#diff-98d01f1091b78761c5134f4b089e514f716b7ecd274b91dcfbb8d1c685ff7ba6R218
```
K_THREAD_DEFINE(producer_thread_id, PRODUCER_STACK_SIZE, producer_thread, NULL, NULL, NULL, 5, 0, 0);
```

rodrigopex commented on samples/subsys/zbus/benchmark/src/benchmark.c:
```diff
@@ -183,7 +190,11 @@ static void producer_thread(void)
 	zbus_chan_finish(&bm_channel);
 
-	uint32_t start = k_uptime_get_32();
+#if defined(CONFIG_ARCH_POSIX)
+	uint32_t start_ns = native_rtc_gettime_us(RTC_CLOCK_PSEUDOHOSTREALTIME) * NSEC_PER_USEC;
+#else
+	uint32_t start_ns = k_cyc_to_ns_near32(sys_clock_cycle_get_32());
+#endif
```
I would not use the declaration inside the ifdef. 

Suggestion:
```c
#if defined(CONFIG_ARCH_POSIX)
#define GET_ARCH_START_TIME (native_rtc_gettime_us(RTC_CLOCK_PSEUDOHOSTREALTIME) * NSEC_PER_USEC)
#else
#define GET_ARCH_START_TIME (k_cyc_to_ns_near32(sys_clock_cycle_get_32()))
#endif

uint32_t start_ns = GET_ARCH_START_TIME;
```

alpsayin replied:
Beautiful.
I called it `GET_ARCH_TIME_NS()` to also imply function calls are being made. That ok?

rodrigopex replied:
Better

rodrigopex commented on samples/subsys/zbus/benchmark/src/benchmark.c:
```diff
@@ -197,20 +208,27 @@ static void producer_thread(void)
 		zbus_chan_notify(&bm_channel, K_MSEC(200));
 	}
-	uint32_t duration = (k_uptime_get_32() - start);
+
+#if defined(CONFIG_ARCH_POSIX)
+	uint32_t end_ns = native_rtc_gettime_us(RTC_CLOCK_PSEUDOHOSTREALTIME) * NSEC_PER_USEC;
+#else
+	uint32_t end_ns = k_cyc_to_ns_near32(sys_clock_cycle_get_32());
+#endif
+
+	uint32_t duration = end_ns - start_ns;
```
Same as the `start_ns` case.

alpsayin replied:
Same as above, defined a single `GET_ARCH_TIME_NS()` and called it in both instances.

rodrigopex reviewed (changes requested):
I would avoid abbreviations for one-time used variables and declare variables inside ifdefs.

alpsayin committed:
samples: subsys: zbus: benchmark: use native posix host rtc for ARCH_POSIX

Because posix emulated execution is significantly faster than posix
emulated timer and benchmark finishes before even the first tick occurs.

Signed-off-by: Alp Sayin <alp.sayin@amd.com>

carlescufi commented:
@alpsayin can you please fix the failing checks?

rodrigopex commented on samples/subsys/zbus/benchmark/src/benchmark.c:
```diff
@@ -11,8 +11,19 @@
+
+#if defined(CONFIG_ARCH_POSIX)
+#include "native_rtc.h"
+#define GET_ARCH_TIME_NS() (native_rtc_gettime_us(RTC_CLOCK_PSEUDOHOSTREALTIME) * NSEC_PER_USEC)
```
@cfriedt Could you please give us your opinion about line 17? Is this a good practice for `native_posix`?

rodrigopex reviewed (commented):
Just waiting for @cfriedt to comment on the native_posix get-time approach.

alpsayin commented:
Hi @carlescufi thanks for showing interest.
I think 1st failure here is a `checkpatch` bug, `checkpatch` either complains line's too long or I'm "unnecessarily" breaking it. Please see in fact our discussion on 2nd line. What should I do?
https://github.com/zephyrproject-rtos/zephyr/actions/runs/3871989306/jobs/6603001449#step:9:31
https://github.com/zephyrproject-rtos/zephyr/pull/53563#discussion_r1064138980
```
K_THREAD_DEFINE(producer_thread_id, PRODUCER_STACK_SIZE, producer_thread, NULL, NULL, NULL, 5, 0, 0);
```

2nd failure I see here is just me being somehow very unlucky:
https://github.com/zephyrproject-rtos/zephyr/actions/runs/3871989309/jobs/6603001548#step:5:54
```
Err:1 http://azure.archive.ubuntu.com/ubuntu focal/universe amd64 ocaml-base-nox amd64 4.08.1-8
  503  Service Unavailable [IP: 20.106.104.242 80]
E: Failed to fetch http://azure.archive.ubuntu.com/ubuntu/pool/universe/o/ocaml/ocaml-base-nox_4.08.1-8_amd64.deb  503  Service Unavailable [IP: 20.106.104.242 80]
```

rodrigopex commented:
> 2nd failure I see here is just me being somehow very unlucky: https://github.com/zephyrproject-rtos/zephyr/actions/runs/3871989309/jobs/6603001548#step:5:54
> 
> ```
> Err:1 http://azure.archive.ubuntu.com/ubuntu focal/universe amd64 ocaml-base-nox amd64 4.08.1-8
>   503  Service Unavailable [IP: 20.106.104.242 80]
> E: Failed to fetch http://azure.archive.ubuntu.com/ubuntu/pool/universe/o/ocaml/ocaml-base-nox_4.08.1-8_amd64.deb  503  Service Unavailable [IP: 20.106.104.242 80]
> ```

@alpsayin 

Let's try it again in some minutes.

carlescufi commented:
> I think 1st failure here is a `checkpatch` bug, `checkpatch` either complains line's too long or I'm "unnecessarily" breaking it. Please see in fact our discussion on 2nd line. What should I do?

I think we can let this one go, but double-checking with @stephanosio

rodrigopex commented on samples/subsys/zbus/benchmark/src/benchmark.c:
```diff
@@ -197,20 +207,24 @@ static void producer_thread(void)
 }
 
-K_THREAD_DEFINE(producer_thread_id, 1024, producer_thread, NULL, NULL, NULL, 5, 0, 0);
+K_THREAD_DEFINE(producer_thread_id, PRODUCER_STACK_SIZE, producer_thread, \
```
@alpsayin remove the `\`. Now I could see that at the end of line 229.

keith-packard commented on samples/subsys/zbus/benchmark/src/benchmark.c:
```diff
@@ -11,8 +11,19 @@
+#include "native_rtc.h"
+#define GET_ARCH_TIME_NS() (native_rtc_gettime_us(RTC_CLOCK_PSEUDOHOSTREALTIME) * NSEC_PER_USEC)
+#else
+#define GET_ARCH_TIME_NS() (k_cyc_to_ns_near32(sys_clock_cycle_get_32()))
```
I was seeing issues on the ARC QEMU platform where the tick count runs at 100Hz; does this API provide a higher resolution value for that case?

alpsayin replied:
Seems so
```
-- west build: running target run_qemu
[0/1] To exit from QEMU enter: 'CTRL+a, x'[QEMU] CPU: arcem
*** Booting Zephyr OS build zephyr-v3.2.0-3158-g7bb6557ba710 ***
I: Benchmark 1 to 8: Dynamic memory, SYNC transmission and message size 256
I: Bytes sent = 262144, received = 262144
I: Average data rate: 2.11MB/s
I: Duration: 118061.400us

@118061400
```
```
[0/1] To exit from QEMU enter: 'CTRL+a, x'[QEMU] CPU: archs
*** Booting Zephyr OS build zephyr-v3.2.0-3158-g7bb6557ba710 ***
I: Benchmark 1 to 8: Dynamic memory, SYNC transmission and message size 256
I: Bytes sent = 262144, received = 262144
I: Average data rate: 2.12MB/s
I: Duration: 117709.400us

@117709400
```
```
[0/1] To exit from QEMU enter: 'CTRL+a, x'[QEMU] CPU: hs6x
*** Booting Zephyr OS build zephyr-v3.2.0-3158-g7bb6557ba710 ***
I: Benchmark 1 to 8: Dynamic memory, SYNC transmission and message size 256
I: Bytes sent = 262144, received = 262144
I: Average data rate: 3.12MB/s
I: Duration: 80011.600us

@80011600
```

stephanosio commented:
> I think we can let this one go, but double-checking with @stephanosio

The package download failure has been fixed and can be ignored. The `LINE_CONTINUATIONS` failure, however, is valid -- it is warning that `\` is not necessary and should be removed.

alpsayin commented:
Of course 🤦🏼 Dont need `\` when inside parenthesis...
Thanks both!

alpsayin commented:
@rodrigopex I have a feeling I should also compare ticks for good measure.
I couldn't trust that cycles isn't going to overflow.
Some numbers I saw are 100s of Mbytes/s whereas arc returned 2-3.
Can we really expect this much deviation?

carlescufi commented:
Merging this now, additional improvements can be done in a subsequent PR.

rodrigopex commented:
> @rodrigopex I have a feeling I should also compare ticks for good measure. I couldn't trust that cycles isn't going to overflow. Some numbers I saw are 100s of Mbytes/s whereas arc returned 2-3. Can we really expect this much deviation?

@alpsayin I need to think more about it. We can discuss it in the following days. Ok?

keith-packard commented:
This doesn't fix the problem on qemu_nios2 where the cycle counter resolution is the the same as the system clock (100Hz by default).  Cranking up the tick rate (`CONFIG_SYS_CLOCK_TICKS_PER_SEC`) to 1kHz fixes the problem.

Output with minimal libc (with 1kHz tick):
```
*** Booting Zephyr OS build zephyr-v3.2.0-3310-g4ec18e5fbf52 ***
I: Benchmark 1 to 8: Dynamic memory, SYNC transmission and message size 256
I: Bytes sent = 262144, received = 262144
I: Average data rate: 7.35MB/s
I: Duration: 34000.0us

@34000000
```

Output with picolibc (with 1kHz tick):
```
*** Booting Zephyr OS build zephyr-v3.2.0-3310-g4ec18e5fbf52 ***
I: Benchmark 1 to 8: Dynamic memory, SYNC transmission and message size 256
I: Bytes sent = 262144, received = 262144
I: Average data rate: 62.50MB/s
I: Duration: 4000.0us

@4000000
```

As you can see, when using picolibc, the test completes in 4ms, well under the 10ms system tick.

You should be able to verify this using this command:
```
$ west build -b qemu_nios2 samples/subsys/zbus/benchmark -T sample.zbus.benchmark_sync -- -DCONFIG_PICOLIBC=y
```

I'll submit another PR that just increases the TICK value for these tests.

rodrigopex commented:
That would be great.

alpsayin commented:
> > @rodrigopex I have a feeling I should also compare ticks for good measure. I couldn't trust that cycles isn't going to overflow. Some numbers I saw are 100s of Mbytes/s whereas arc returned 2-3. Can we really expect this much deviation?
> 
> @alpsayin I need to think more about it. We can discuss it in the following days. Ok?

I checked, and we get away with (mostly by chance). In the slowest platforms, tests still finish before cycles can roll over. 
It also seems nanoseconds was a bit of an overkill but oh well  `¯\_(ツ)_/¯`
